### PR TITLE
Handling failed progressions within as

### PR DESCRIPTION
**Pull request by chrisdugne**
## Detailed purpose of the PR

Finished progressions where remove from Async storage.
Finished progression in `failure` or `extralifeRefused` are required for hero algo.

---> This PR leaves all progressions within AS and register the synced one in an array not to sync then twice.

### Testing strategy

- [x] I ensured my code is tested: 
- [x] I took care to mention breaking changes and extra libs if appropriate

### CODE REVIEW

**Comment by chrisdugne** [comment]
[[7] [Hero] Algo pour parcourrir les complétions et retourner le contenu du hero](https://trello.com/c/1TGa42xh/1424-7-hero-algo-pour-parcourrir-les-compl%C3%A9tions-et-retourner-le-contenu-du-hero)

File: src/redux/actions/progressions/synchronize.js [hunk 1]
```diff
@@ -63,18 +68,25 @@ export const synchronizeProgressions: StoreAction<Action> = async (dispatch, get
   const {services} = options;
 
   const progressions = await services.Progressions.getAll();
+  const _synchronizedProgressions: string = await AsyncStorage.getItem(SYNCHRONIZED_PROGRESSIONS);
```
Comment (VincentCATILLON) [inline src/redux/actions/progressions/synchronize.js]: use datalayer for this and test it so :)

File: src/redux/actions/progressions/synchronize.js [hunk 2]
```diff
@@ -63,18 +68,25 @@ export const synchronizeProgressions: StoreAction<Action> = async (dispatch, get
...
+  const synchronizedProgressions = JSON.parse(_synchronizedProgressions) || [];
 
   await pMap(
-    sortProgressionChronologicaly(progressions).filter(isDone),
+    sortProgressionChronologicaly(progressions).filter(
+      p => isDone(p) && !synchronizedProgressions.includes(p._id)
```
Comment (VincentCATILLON) [inline src/redux/actions/progressions/synchronize.js]: can you please define a function for this thing ? ie: `!isAlreadySynchronized()` as described in analysis ?

File: src/redux/actions/progressions/synchronize.js [hunk 3]
```diff
@@ -63,18 +68,25 @@ export const synchronizeProgressions: StoreAction<Action> = async (dispatch, get
...
       if (_id) {
+        synchronizedProgressions.push(_id);
         return dispatch(synchronizeProgression(_id));
       }
       return Promise.resolve();
     },
     {concurrency: 1}
   );
 
+  await AsyncStorage.setItem(SYNCHRONIZED_PROGRESSIONS, JSON.stringify(synchronizedProgressions));
```
Comment (VincentCATILLON) [inline src/redux/actions/progressions/synchronize.js]: should be a function in data layer (tested)

File: src/redux/actions/progressions/synchronize.test.js [hunk 4]
```diff
@@ -101,9 +108,55 @@ describe('Progressions synchronization', () => {
       // $FlowFixMe
       const actual = await synchronizeProgressions(store.dispatch, store.getState, {services});
 
+      expect(AsyncStorage.setItem).toHaveBeenCalledTimes(1);
+      expect(AsyncStorage.setItem).toHaveBeenCalledWith(
+        SYNCHRONIZED_PROGRESSIONS,
+        JSON.stringify([failureProgression._id, successProgression._id])
+      );
```
Comment (VincentCATILLON) [inline src/redux/actions/progressions/synchronize.test.js]: you can mock the datalayer file so :)

File: src/redux/actions/progressions/synchronize.js (hunk 1 above)
Comment (VincentCATILLON) [reply src/redux/actions/progressions/synchronize.js]: btw, it's not a string, it's an array of strings

File: src/redux/actions/progressions/synchronize.js (hunk 2 above)
Comment (chrisdugne) [reply src/redux/actions/progressions/synchronize.js]: not that much of an util like isDone though, dependending on `datalayer.synchronizedProgressions`

**Comment by VincentCATILLON** [comment]
Merging it because it's too hard to work with 2 branches / PR

**Pull request merged**